File: data/src/main/java/com/neuifo/data/converter/ResponseConverterFactory.java
package com.neuifo.data.converter;

import com.google.gson.Gson;
import com.google.gson.TypeAdapter;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonWriter;
import com.neuifo.data.domain.utils.ClassConstructorLoader;

import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.lang.annotation.Annotation;
import java.lang.reflect.Constructor;
import java.lang.reflect.Type;
import java.nio.charset.Charset;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okio.Buffer;
import retrofit2.Converter;
import retrofit2.Retrofit;

/**
 * Created by neuifo on 16/7/25.
 * 处理响应结果
 */
public class ResponseConverterFactory<T extends Converter> extends Converter.Factory {

    public static ResponseConverterFactory create(Gson gson, Class tClass) {
        return new ResponseConverterFactory(gson, tClass);
    }

    private final Gson gson;
    private final Class clazz;

    private ResponseConverterFactory(Gson gson, Class<T> tClass) {
        this.gson = gson;
        this.clazz = tClass;
    }


    @Override
    public Converter<ResponseBody, ?> responseBodyConverter(Type type, Annotation[] annotations, Retrofit retrofit) {
        TypeAdapter<?> adapter = gson.getAdapter(TypeToken.get(type));
        Object[] objArr = new Object[]{gson, adapter, type};
        try {
            Constructor<T> constructor = ClassConstructorLoader.getAppropriateConstructor(clazz, objArr);
            constructor.setAccessible(true);
            T t = constructor.newInstance(objArr);
            return t;
        } catch (Exception e) {
        }
        return null;
    }

    @Override
    public Converter<?, RequestBody> requestBodyConverter(Type type,
                                                          Annotation[] parameterAnnotations, Annotation[] methodAnnotations, Retrofit retrofit) {
        TypeAdapter<?> adapter = gson.getAdapter(TypeToken.get(type));
        return new GsonRequestBodyConverter<>(gson, adapter);
    }


    static class GsonRequestBodyConverter<T> implements Converter<T, RequestBody> {

        private static final MediaType MEDIA_TYPE = MediaType.parse("application/json; charset=UTF-8");
        private static final Charset UTF_8 = Charset.forName("UTF-8");

        private final Gson gson;
        private final TypeAdapter<T> adapter;

        GsonRequestBodyConverter(Gson gson, TypeAdapter<T> adapter) {
            this.gson = gson;
            this.adapter = adapter;
        }

        @Override
        public RequestBody convert(T value) throws IOException {
            Buffer buffer = new Buffer();
            Writer writer = new OutputStreamWriter(buffer.outputStream(), UTF_8);
            JsonWriter jsonWriter = gson.newJsonWriter(writer);
            adapter.write(jsonWriter, value);
            jsonWriter.close();
            return RequestBody.create(MEDIA_TYPE, buffer.readByteString());
        }

    }
}

File: data/src/main/java/com/neuifo/data/converter/ResponseConverterFactory.kt
package com.neuifo.data.converter

import com.google.gson.Gson
import com.google.gson.TypeAdapter
import com.google.gson.reflect.TypeToken
import com.neuifo.data.domain.utils.ClassConstructorLoader.getAppropriateConstructor
import okhttp3.MediaType
import okhttp3.RequestBody
import okhttp3.ResponseBody
import okio.Buffer
import retrofit2.Converter
import retrofit2.Retrofit
import java.io.IOException
import java.io.OutputStreamWriter
import java.io.Writer
import java.lang.reflect.Constructor
import java.lang.reflect.Type
import java.nio.charset.Charset

/**
 * Created by neuifo on 16/7/25.
 * 处理响应结果
 */
class ResponseConverterFactory<T : Converter<*, *>> private constructor(
    private val gson: Gson,
    private val clazz: Class<T>
) : Converter.Factory() {
    override fun responseBodyConverter(
        type: Type,
        annotations: Array<Annotation>,
        retrofit: Retrofit
    ): Converter<ResponseBody, T>? {
        val adapter =
            gson.getAdapter(TypeToken.get(type))
        val objArr = arrayOf<Any?>(gson, adapter, type)
        try {
            val constructor: Constructor<T> =
                getAppropriateConstructor<T>(clazz, objArr)
            constructor.isAccessible = true
            return constructor.newInstance(*objArr) as Converter<ResponseBody, T>
        } catch (e: Exception) {
        }
        return null
    }

    override fun requestBodyConverter(
        type: Type,
        parameterAnnotations: Array<Annotation>,
        methodAnnotations: Array<Annotation>,
        retrofit: Retrofit
    ): Converter<*, RequestBody> {
        val adapter =
            gson.getAdapter(TypeToken.get(type))
        return GsonRequestBodyConverter(
            gson,
            adapter
        )
    }

    internal class GsonRequestBodyConverter<T>(
        private val gson: Gson,
        private val adapter: TypeAdapter<T>
    ) : Converter<T, RequestBody> {
        @Throws(IOException::class)
        override fun convert(value: T): RequestBody {
            val buffer = Buffer()
            val writer: Writer = OutputStreamWriter(
                buffer.outputStream(),
                UTF_8
            )
            val jsonWriter = gson.newJsonWriter(writer)
            adapter.write(jsonWriter, value)
            jsonWriter.close()
            return RequestBody.create(
                MEDIA_TYPE,
                buffer.readByteString()
            )
        }

        companion object {
            private val MEDIA_TYPE =
                MediaType.parse("application/json; charset=UTF-8")
            private val UTF_8 = Charset.forName("UTF-8")
        }

    }

    companion object {
        @JvmStatic
        fun create(
            gson: Gson,
            tClass: Class<out Converter<*,*>>
        ): ResponseConverterFactory<*> {
            return ResponseConverterFactory(gson, tClass)
        }
    }

}